_posts/2023-05-06-reverse-engineering-gba.markdown

```diff
@@ -0,0 +1,1064 @@
+
+You can use the `n` (`next`) command to execute the current instruction and pause on the next instruction.
+>![](/assets/img/reverse-engineering/debugger-next.png)<br>
+>Executing the next instruction while debugging
+
+It is also possible to execute multiple instructions using the `trace <count>` command. For example, `trace 5` is the equivalent of using the `n` command 5 times. Note that `trace` prints condensed game state views for each intermediate instruction to save space.
+>![](/assets/img/reverse-engineering/debugger-trace.png)<br>
+>Tracing multiple instructions at once
+
+To exit debugging, use the `c` (`continue`) command. This will cause the game to resume.
+
+If you want to execute the same command you just executed, you can hit `Enter` with no command entered. This can be useful for commands like `n`. mGBA also keeps a history of executed commands, which you can access with the up/down arrows similar to a regular terminal window.
+
+#### Reading/writing memory
+The memory viewer can be used while the debugger has paused the game, though it is sometimes convenient to read/write memory values directly with the debugger console.
+
+The read command uses the format `r/<size> <address>`, where `<size>` is 1, 2, or 4 bytes. For example, to read the player HP value from earlier, use the command `r/1 0x200419E`. The address can be expressed in either decimal or hex; the standard `0x` prefix denotes a hex value.
+>![](/assets/img/reverse-engineering/debugger-read.png)<br>
+>Reading a memory value
+
+Note that most debugger console commands will pause the game if it is not already paused.
+
+Writing values to memory uses the format `w/<size> <address> <write value>`. For instance, the command `w/1 0x200419E 20` will set the player's HP to 20. Like the address, the write value can be either decimal or hex. There is also a `w/r` command to write a value to a register instead of memory.
+
+#### Breakpoints
+Similar to debuggers in higher-level languages, a breakpoint pauses program execution if a specific line of code is reached. This has several uses, like figuring out whether a line of code is reached when performing an action in-game, or stopping program execution at a specific function to debug it.
+
+A breakpoint can be set using the `b` (`break`) command, with the format `b <address>`. As a demonstration, set a breakpoint at the beginning of the function `FUN_08044b28` using `b 0x08044b28`. This will pause game execution, so unpause the game with the `c` command.
+>![](/assets/img/reverse-engineering/debugger-add-breakpoint.png)<br>
+>Adding a breakpoint
+
+With the breakpoint added, try to move around in-game. This will execute `FUN_08044b28` and hit the breakpoint, pausing the game's execution.
```

Comment:
```suggestion
With the breakpoint added, try to moving your character around in the gba emulator. This will execute `FUN_08044b28` and hit the breakpoint, pausing the game's execution.
```

Reply:
The term "emulator" includes both the game window and other supporting windows like the debugging console, so "in-game" is a more narrow description. I'll mention the character though.